### PR TITLE
Kubernetes client openapi watcher

### PR DESCRIPTION
**Pull request by msupic**
@melix - could you please review gradle related code

This PR includes the following modifications:
- moved the kubernetes client gradle plugin from `buildSrc` to a new module `build-logic`
- created a new gradle task `createWatcherOpenApiSpec` (kubernetes client gradle plugin) which modifies the already downloaded openapi spec so it can be used for creating api for streaming. The task:
  - removes `components` section (since the model is created by another gradle task in another micronaut module) 
  - filters out all operations which are not `get` and which doesn't contain the `watch` param
  - modifies response schemas of the kept operations so each schema references a list of items instead of list object (for example, list of `V1Namespace` instead of `V1NamespaceList`)
  - creates a type mapping file which contains mappings between response types from the new spec file and generic types (for example, mapping between `V1Namespace` and `WatchEvent<V1Namespace>`. The mapping file is later used by the micronaut openapi generator to create something like:
  ```
    @Get("/api/v1/namespaces")
    @Consumes({"application/json", "application/yaml", "application/vnd.kubernetes.protobuf", "application/json;stream=watch", "application/vnd.kubernetes.protobuf;stream=watch"})
    Flux<WatchEvent<V1Namespace>> listNamespace(
        @QueryValue("pretty") @Nullable String pretty,
    ```
    The generic type is needed since all kubernetes server streamed events have the same wrapper. For example,
    ```
         {"type":"ADDED","object":{"kind":"Namespace","apiVersion":"v1",...}}
         {"type":"MODIFIED","object":{"kind":"Secret","apiVersion":"v1",...}}
         {"type":"DELETED","object":{"kind":"Pod","apiVersion":"v1",...}}
    ```
- created a custom json deserializer `WatchEventDeserializer` to deserialize streamed events to `WatchEvent`
```
public record WatchEvent<T>(
    @NonNull String type,
    @Nullable T object,
    @Nullable V1Status status
```
- modified kubernetes http client to use managed `MessageBodyHandlerRegistry` and `MediaTypeCodecRegistry`
- updated docs

### CODE REVIEW

File: kubernetes-client-openapi-watcher/build.gradle [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+plugins {
+    id 'io.micronaut.build.internal.kubernetes-module'
+    id 'io.micronaut.openapi' version libs.versions.micronaut.gradle.plugin
+    id 'io.micronaut.kubernetes.client.openapi'
+}
+
+micronautBuild {
+    binaryCompatibility {
+        enabled.set(false)
```
Comment (melix) [inline kubernetes-client-openapi-watcher/build.gradle]: :x: Instead of disabling, you can use `enabledAfter("...")`

**Review by melix — APPROVED** [review]

File: kubernetes-client-openapi-watcher/build.gradle (hunk 1 above)
Comment (msupic) [reply kubernetes-client-openapi-watcher/build.gradle]: Modified

File: build-logic/src/main/java/io/micronaut/kubernetes/client/openapi/tasks/CreateWatcherSpec.java [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+        return (Map<String, Object>) map.get(key);
+    }
+
+    private Yaml createYaml() {
+        LoaderOptions loaderOptions = new LoaderOptions();
+        loaderOptions.setCodePointLimit(10485760);
+        DumperOptions dumperOptions = new DumperOptions();
+        dumperOptions.setPrettyFlow(true);
+        dumperOptions.setDefaultFlowStyle(DumperOptions.FlowStyle.BLOCK);
+        return new Yaml(new Constructor(loaderOptions), new Representer(dumperOptions));
```
Comment (graemerocher) [inline build-logic/src/main/java/io/micronaut/kubernetes/client/openapi/tasks/CreateWatcherSpec.java]: use `SafeConstructor` instead

File: kubernetes-client-openapi-watcher/src/main/java/io/micronaut/kubernetes/client/openapi/watcher/WatchEventDeserializer.java [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+ * <p>Custom deserializer for kubernetes streamed events.</p>
+ * <br />
+ * <p>The deserializer first checks a value of the {@code type} field to determine how to
+ * deserialize a value of the {@code object} field. If the value of the {@code type} field is equal
+ * to {@code ERROR}, then the {@code object} field will be deserialized into an instance of {@link V1Status},
+ * otherwise the {@code object} field will be deserialized into an instance of the class passed as the type parameter.
+ * </p>
+ */
+@Singleton
+class WatchEventDeserializer implements Deserializer<WatchEvent<?>> {
```
Comment (graemerocher) [inline kubernetes-client-openapi-watcher/src/main/java/io/micronaut/kubernetes/client/openapi/watcher/WatchEventDeserializer.java]: can be `final`

Comment (graemerocher) [inline kubernetes-client-openapi-watcher/src/main/java/io/micronaut/kubernetes/client/openapi/watcher/WatchEventDeserializer.java]: annotate with `@Internal` for now

File: src/main/docs/guide/kubernetes-client-openapi-watcher.adoc [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+dependency:io.micronaut.kubernetes:micronaut-kubernetes-client-openapi-watcher[]
+
+Then you can simply use Micronaut injection to get configured apis object from package `io.micronaut.kubernetes.client.openapi.watcher.api`:
+
+snippet::micronaut.client.SecretController[project-base="examples/example-kubernetes-client-openapi-watcher", source="main"]
+
+[IMPORTANT]
+The streaming will work only if the `watch` parameter is set to `true`.
+
+The Kubernetes streamed events are deserialized into the following `io.micronaut.kubernetes.client.openapi.watcher.WatchEvent` class:
```
Comment (graemerocher) [inline src/main/docs/guide/kubernetes-client-openapi-watcher.adoc]: use `api:` macro to link to the `WatchEvent` javadoc

File: src/main/docs/guide/kubernetes-client-openapi-watcher.adoc [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+
+[source,java]
+----
+import io.micronaut.core.annotation.NonNull;
+import io.micronaut.core.annotation.Nullable;
+import io.micronaut.kubernetes.client.openapi.model.V1Status;
+import io.micronaut.serde.annotation.Serdeable;
+
+@Serdeable
+public record WatchEvent<T>(
```
Comment (graemerocher) [inline src/main/docs/guide/kubernetes-client-openapi-watcher.adoc]: you can probably use an ascidoc include to induce the actual source

**Review by graemerocher — CHANGES_REQUESTED** [review]

File: build-logic/src/main/java/io/micronaut/kubernetes/client/openapi/tasks/CreateWatcherSpec.java (hunk 2 above)
Comment (msupic) [reply build-logic/src/main/java/io/micronaut/kubernetes/client/openapi/tasks/CreateWatcherSpec.java]: Modified

File: kubernetes-client-openapi-watcher/src/main/java/io/micronaut/kubernetes/client/openapi/watcher/WatchEventDeserializer.java (hunk 3 above)
Comment (msupic) [reply kubernetes-client-openapi-watcher/src/main/java/io/micronaut/kubernetes/client/openapi/watcher/WatchEventDeserializer.java]: Added

Comment (msupic) [reply kubernetes-client-openapi-watcher/src/main/java/io/micronaut/kubernetes/client/openapi/watcher/WatchEventDeserializer.java]: Added

File: src/main/docs/guide/kubernetes-client-openapi-watcher.adoc (hunk 4 above)
Comment (msupic) [reply src/main/docs/guide/kubernetes-client-openapi-watcher.adoc]: Modified

File: src/main/docs/guide/kubernetes-client-openapi-watcher.adoc (hunk 5 above)
Comment (msupic) [reply src/main/docs/guide/kubernetes-client-openapi-watcher.adoc]: Modified

**Review by graemerocher — APPROVED** [review]

**Pull request merged**